servlet/src/main/java/io/grpc/servlet/ServletServerStream.java

```diff
@@ -0,0 +1,342 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.servlet;
+
+import static io.grpc.internal.GrpcUtil.CONTENT_TYPE_GRPC;
+import static io.grpc.internal.GrpcUtil.CONTENT_TYPE_KEY;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.util.logging.Level.FINE;
+import static java.util.logging.Level.FINEST;
+import static java.util.logging.Level.WARNING;
+
+import com.google.common.io.BaseEncoding;
+import com.google.common.util.concurrent.MoreExecutors;
+import io.grpc.Attributes;
+import io.grpc.InternalLogId;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.Status.Code;
+import io.grpc.internal.AbstractServerStream;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.internal.SerializingExecutor;
+import io.grpc.internal.StatsTraceContext;
+import io.grpc.internal.TransportFrameUtil;
+import io.grpc.internal.TransportTracer;
+import io.grpc.internal.WritableBuffer;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.servlet.AsyncContext;
+import javax.servlet.WriteListener;
+import javax.servlet.http.HttpServletResponse;
+
+final class ServletServerStream extends AbstractServerStream {
+
+  private static final Logger logger = Logger.getLogger(ServletServerStream.class.getName());
+
+  private final ServletTransportState transportState;
+  private final Sink sink = new Sink();
+  private final AsyncContext asyncCtx;
+  private final HttpServletResponse resp;
+  private final Attributes attributes;
+  private final String authority;
+  private final InternalLogId logId;
+  private final AsyncServletOutputStreamWriter writer;
+
+  ServletServerStream(
+      AsyncContext asyncCtx,
+      StatsTraceContext statsTraceCtx,
+      int maxInboundMessageSize,
+      Attributes attributes,
+      String authority,
+      InternalLogId logId) throws IOException {
+    super(ByteArrayWritableBuffer::new, statsTraceCtx);
+    transportState =
+        new ServletTransportState(maxInboundMessageSize, statsTraceCtx, new TransportTracer());
+    this.attributes = attributes;
+    this.authority = authority;
+    this.logId = logId;
+    this.asyncCtx = asyncCtx;
+    this.resp = (HttpServletResponse) asyncCtx.getResponse();
+    resp.getOutputStream().setWriteListener(new GrpcWriteListener());
```

Comment:
When testing against Payara server, this class seemingly fails to completely initialise unless the `writer` field is initialised before the write listener on this line. When swapped I get NPEs in the OutputBuffer, but I'm willing to assume this could just be an expected Grizzly bug. I do, however, at least get the expected communication between the client and server.

Reply:
In https://github.com/grpc/grpc-java/pull/8596/commits/c903d75c1aca0dd29406282e20e1a0b37759b00b#diff-61714cd00cc4b70a569fd5ac7ac6919bcc9e2e76b6c0384ffbbeda9587c5ff93 of #8596, I swapped the assignment of the `writer` field. Thanks for catching this.